pipeline/minimmit.md

```diff
@@ -58,134 +60,138 @@ fn verify(c, c') -> bool;
 
 ## 7. Helpers
 
+_In every helper function, the first parameter `r` is the replica whose local state the helper inspects or updates._
+
 ```text
-// Find a valid parent to build on
-fn select_parent(v) -> (c', v') {
+// Replica `r` selects a valid parent to build on
+fn select_parent(r, v) -> (c', v') {
     let i = v - 1;
     while i >= 0 {
-        if notarization(c', i) ∈ proofs[i] {
-            // If there are multiple, pick any.
-            return (c', i);
+        if notarization(c', i) ∈ r.proofs[i] {
+            return (c', i); // If there are multiple, pick any
         }
-        if nullification(i) ∈ proofs[i] {
+        if nullification(i) ∈ r.proofs[i] {
             i -= 1;
             continue;
         }
-        return ⊥;
+        return (⊥, ⊥); // No proofs for view i, cannot proceed
     }
-    return genesis;
+    return (genesis, ⊥);
 }
 
-// Ensure there are proofs for all views between `v` and `v'`
-fn valid_parent(v, (c', v')) -> bool {
+// Replica `r` ensures there are nullifications for all views `[v', v)` and that there is a `notarization(c', v')`
+fn valid_parent(r, v, (c', v')) -> bool {
     let i = v - 1;
     while i > v' {
-        if nullification(i) ∈ proofs[i] {
-            i -= 1;
-            continue;
+        if nullification(i) ∉ r.proofs[i] {
+            return false;
         }
-        return false;
+        i -= 1;
     }
-    return notarization(c', v') ∈ proofs[v']
+    return notarization(c', v') ∈ r.proofs[v']
 }
 
-// Enter view `next`
-fn enter_view(next) {
-    if view >= next {
+// Replica `r` enters view `next` if greater than the current view
+fn enter_view(r, next) {
+    if r.view >= next {
         return;
     }
-    view = next;
-    notarized = ⊥;
-    nullified = false;
-    timer = 2Δ;
+    r.view = next;
+    r.notarized = ⊥;
+    r.nullified = false;
+    r.timer = 2Δ;
 }
 
-// Record a message from a `replica`
-fn record_message(replica, message) -> bool {
-    if replica ∉ messages[message.view] {
-        messages[message.view][replica] = [];
+// Replica `r` records a message, `m`, received from a replica `r'`
+fn record_message(r, r', m) -> bool {
+    if m.v ∉ r.messages {
```

Comment:
I left these out before because I assumed they were "implicit" (I totally buy including them to make it clearer).